### DISCUSSION TITLE
Running in GPU instance

### DISCUSSION THREAD

**Discussion by sankagirirajeev** [microsoft/presidio#1185] [Q&A]
I'm running Persidio on around 60 lakh records. It takes more than 4 minutes to process 1000k data. Is there any way to run it in the fattest way possible?

**Comment by omri374** [comment]
Have you looked at the batch option? Presidio has a `BatchAnalyzerEngine` object, see [sample here](https://microsoft.github.io/presidio/samples/python/batch_processing/). The underlying NER model also affects the performance, and each model has different requirements, latency, and might be faster with a GPU.

Which model are you using? Please also share the code, as 4 minutes for 1mb sounds too much. Are you initializing the `AnalyzerEngine` multiple times?

**Comment by sankagirirajeev** [comment]
```py
from presidio_analyzer import AnalyzerEngine
from presidio_anonymizer import AnonymizerEngine
from presidio_anonymizer.entities import RecognizerResult, OperatorConfig
import datetime
import pandas as pd
import spacy
import re
spacy.require_gpu()
analyzer = AnalyzerEngine()
anonymizer = AnonymizerEngine()
data_sample =pd.read_parquet('xxx.parquet')

for index,row in data_sample.iterrows():
    content = row['content']
   
    if len(content) < 1000000:
    
        results = analyzer.analyze(text=content,
                               
                               language='en')
        """for result in results:
            print("Type:", result.entity_type)
            #print("Value:", result.text)
            #print("Score:", result.score)
            print(result)"""
```

**Comment by sankagirirajeev** [comment]
I used batchanalyzer
start = timer()
analyzer_results = batch_analyzer.analyze_iterator(data_sample_content, language='en')
analyzer_results = list(analyzer_results)
print(timer()-start)


its taking same amount time

**Reply by omri374** [reply]
I believe that the `en_core_web_lg` model is CPU optimized, but I'm not sure.

**Comment by sankagirirajeev** [comment]
is any other way we can process the data?

**Reply by omri374** [reply]
It requires some profiling to understand the bottlenecks. The `BatchAnalyzerEngine` leverages a capability for multi-core processing in spaCy. Perhaps try remove the `spacy.require_gpu()` flag.

**Reply by sankagirirajeev** [reply]
taking same timing

**Reply by sankagirirajeev** [reply]
it would be more helpful ..if i find a  solution? need to run large dataset within a day

**Reply by omri374** [reply]
Have you ran any profiling? Snakeviz is a nice option: https://jiffyclub.github.io/snakeviz/

**Reply by omri374** [reply]
We have deployment examples in [Pyspark (using Python)](https://microsoft.github.io/presidio/samples/deployments/spark/) and [Kubernetes (using Docker)](https://microsoft.github.io/presidio/samples/deployments/k8s/). This would likely be the next step.

**Reply by sankagirirajeev** [reply]
Is this code useful for speeding up processing?

**Reply by omri374** [reply]
It allows you to run Presidio in parallel, using Spark or Kubernetes. Profiling is useful if you're interested in understanding where the bottleneck is.

**Comment by bafflejs** [comment]
@omri374 Is there a way to run presidio on a gpu? I want to run certain models with gpu, how do I do that with presidio?

**Reply by omri374** [reply]
@bafflejs it depends on which NLP model you're running. 
To use spaCy with GPU, run `pip install -U spacy[cuda113]` with your cuda version. For transformers
For transformers and other torch based models, it's likely that you'd have to install torch in a GPU enabled way.

For GLiNER, you should specify the `map_location`
https://github.com/microsoft/presidio/blob/a78e659f3df6a5e890d8a57558dc2ae5a5f474de/presidio-analyzer/presidio_analyzer/predefined_recognizers/gliner_recognizer.py#L37

**Reply by bafflejs** [reply]
Thanks a lot! I also tried the similar approach, but didn't see difference in inference time. Will try with different cuda version again.

**Reply by shubh3ai** [reply]
@omri374 I have torch with GPU support installed and I can use HF pipeline on GPU but Presidio always loads the transformers model on CPU.

**Reply by shubh3ai** [reply]
I found the issue. Using transformers on GPUs requires Spacy with GPU installed. So even though, HF can run a model on GPU, Spacy will not be able to use it. 

While initializing a `hf_token_pipe`, spacy finds the device using 
```py
    try:
        device = get_torch_default_device().index
        if device is None:
            device = -1
    except Exception:
        device = -1
```

I think the way Spacy and transformers are intertwined is bound to trip people up.

**Reply by omri374** [reply]
Thanks @shubh3ai, when you meniton that GPUs require spacy with GPU installed, do you mean this type of installation `pip install -U spacy[cuda113]` or something else?

We will look into creating documentation specifically for GPUs.

**Reply by omri374** [reply]
see #1566

**Reply by alan-insam** [reply]
> see #1566
Hey, did you manage to find out how to use the GPU in the end? Any recommended solution?

**Reply by alan-insam** [reply]
> I found the issue. Using transformers on GPUs requires Spacy with GPU installed. So even though, HF can run a model on GPU, Spacy will not be able to use it.
> 
> While initializing a `hf_token_pipe`, spacy finds the device using
> 
> ```python
>     try:
>         device = get_torch_default_device().index
>         if device is None:
>             device = -1
>     except Exception:
>         device = -1
> ```
> 
> I think the way Spacy and transformers are intertwined is bound to trip people up.

Hey, did you found a solution?

**Comment by rahulgitrepository** [comment]
Is there any documentation regarding how to use transformer model of spaCy within Presidio on a GPU?
